biogeochem/EDCanopyStructureMod.F90

```diff
@@ -469,98 +485,150 @@ subroutine DemoteFromLayer(currentSite,currentPatch,i_lyr)
          ! and adjust things if the demoted area wants to be greater than
-         ! what is available.  sumweights_old is used to sum up the exclusion
-         ! area of those cohorts that have not reached their total canopy area 
-         ! yet.
-         
+         ! what is available. The math is too hard to explain here, see
+         ! the tech note section on promotion/demotion.
+
          if (ED_val_comp_excln .ge. 0.0_r8 ) then
             
-            remainder_area = 0.0_r8
-            sumweights_old = 0.0_r8
+            scale_factor_min  = 1.e10_r8
+            scale_factor      = 0._r8
             currentCohort => currentPatch%tallest
             do while (associated(currentCohort))
 
                if(currentCohort%canopy_layer  ==  i_lyr) then 
-                  currentCohort%excl_weight = demote_area*currentCohort%excl_weight/sumweights
-                  if( currentCohort%excl_weight > currentCohort%c_area ) then
-                     remainder_area = remainder_area + currentCohort%excl_weight-currentCohort%c_area
-                     currentCohort%excl_weight = currentCohort%c_area
-                  else
-                     sumweights_old = sumweights_old + currentCohort%excl_weight
-                  end if
+
+                  currentCohort%excl_weight = currentCohort%excl_weight/sumweights
+                  if( 1._r8/currentCohort%excl_weight <  scale_factor_min )  &
+                        scale_factor_min = 1._r8/currentCohort%excl_weight
+                  
+                  scale_factor = scale_factor + currentCohort%excl_weight * currentCohort%c_area
+
                endif
                currentCohort => currentCohort%shorter      
             enddo
 
-            exceedance_counter = 0
-            do while(remainder_area/demote_area > area_target_precision )
-               
-               ! Keep attempting to add exceedance to members that have not
-               ! lost more area than they started with..
-               
-               sumweights          = 0.0_r8
-               remainder_area_hold = remainder_area
+            ! This is the factor by which we need to multiply
+            ! the demotion probabilities, so the sum result equals
+            ! the total amount to demote
+
+            scale_factor = demote_area/scale_factor
+            
+            if(scale_factor <= scale_factor_min) then
+
+               ! Trivial case, all of the demotion fractions are less than 1.
+
                currentCohort => currentPatch%tallest
-               do while (associated(currentCohort))      
-                  if(currentCohort%canopy_layer == i_lyr)then
-                     if ( currentCohort%excl_weight < (currentCohort%c_area-nearzero) ) then
+               do while (associated(currentCohort))
+                  if(currentCohort%canopy_layer  ==  i_lyr) then 
+                     currentCohort%excl_weight = currentCohort%c_area * currentCohort%excl_weight * scale_factor
```

Comment:
Note to self: ```scale_factor``` includes ```c_area``` so needs to be cancelled out.

Reply:
Update: See comment on line 574 below.  The intent appears to be to compute the excluded area immediately here, which makes sense looking at the debug below and the ```ed_cohort_type``` comments.